Add support for FULL OUTER hash join

mbasmanova opened the pull request.

oerling commented:
Do you need the switch around HashProbe.cpp:56? All values should be supported now?

 

It would probably be possible to define isLeftJoin and isRightJoin to be also true of full oj. Would this make it clearer? Could be either way. Another way of naming these would be isLeft/RightOptional. That is, right is optional for left join. The way it is now is also OK.

 

In comment of setProbedFlag, it may be appropriate to say why an element of ‘rows’ can be null. I suppose this is for a probe key with null/no match.

 

Accepted the diff.

mbasmanova commented:
@oerling Orri, thank you for review.

> Do you need the switch around HashProbe.cpp:56? All values should be supported now? 

Removed.

> It would probably be possible to define isLeftJoin and isRightJoin to be also true of full oj. Would this make it clearer? Could be either way. Another way of naming these would be isLeft/RightOptional. That is, right is optional for left join. The way it is now is also OK. 

I though about it, but decided that it won't be clearer, rather would introduce more terms to think about. I'm keeping this as is.

> In comment of setProbedFlag, it may be appropriate to say why an element of ‘rows’ can be null. I suppose this is for a probe key with null/no match. 

I updated the comment:

```
  /// Sets 'probed' flag for the specified rows. Used by the right and full join
  /// to mark build-side rows that matches join condition. 'rows' may contain
  /// duplicate entries for the cases where single probe row matched multiple
  /// build rows. In case of the full join, 'rows' may include null entries that
  /// correspond to probe rows with no match.
  void setProbedFlag(char** rows, int32_t numRows);
```

aditi-pandit reviewed (commented):
Thanks for this change.
